src/geom/SkyWcs.cc

```diff
@@ -94,181 +170,328 @@ Angle SkyWcs::getPixelScale(Point2D const& pixel) const {
+    return _linearizeSkyToPixel(skyToPixel(coord), coord, skyUnit);
 }
 
-/**
-Compute the pixel position from the sky position
-*/
-std::pair<double, double> SkyWcs::skyToPixel(Angle const& ra, Angle const& dec) const {
-    auto pixel = applyInverse(coord::IcrsCoord(ra, dec));
-    return std::pair<double, double>(pixel[0], pixel[1]);
-};
+AffineTransform SkyWcs::linearizeSkyToPixel(Point2D const& pix, AngleUnit const& skyUnit) const {
+    return _linearizeSkyToPixel(pix, pixelToSky(pix), skyUnit);
+}
 
 std::string SkyWcs::getShortClassName() { return "SkyWcs"; };
 
-SkyWcs SkyWcs::readStream(std::istream& is) { return detail::readStream<SkyWcs>(is); }
+bool SkyWcs::isFlipped() const {
+    double det = getCdMatrix().determinant();
+    if (det == 0) {
+        throw(LSST_EXCEPT(pex::exceptions::RuntimeError, "CD matrix is singular"));
+    }
+    return (det > 0);
+}
+
+std::shared_ptr<SkyWcs> SkyWcs::readStream(std::istream& is) {
+    int version;
+    is >> version;
+    if (version != 1) {
+        throw LSST_EXCEPT(pex::exceptions::InvalidParameterError,
+                          "Unsupported version " + std::to_string(version));
+    }
+    std::string shortClassName;
+    is >> shortClassName;
+    if (shortClassName != SkyWcs::getShortClassName()) {
+        std::ostringstream os;
+        os << "Class name in stream " << shortClassName << " != " << SkyWcs::getShortClassName();
+        throw LSST_EXCEPT(pex::exceptions::InvalidParameterError, os.str());
+    }
+    bool hasFitsApproximation;
+    is >> hasFitsApproximation;
+    auto astStream = ast::Stream(&is, nullptr);
+    auto astObjectPtr = ast::Channel(astStream).read();
+    auto frameSet = std::dynamic_pointer_cast<ast::FrameSet>(astObjectPtr);
+    if (!frameSet) {
+        std::ostringstream os;
+        os << "The AST serialization was read as a " << astObjectPtr->getClassName()
+           << " instead of a FrameSet";
+        throw LSST_EXCEPT(pex::exceptions::InvalidParameterError, os.str());
+    }
+    ast::FrameDict frameDict(*frameSet);
+    return std::make_shared<SkyWcs>(frameDict);
+}
 
-SkyWcs SkyWcs::readString(std::string& str) {
+std::shared_ptr<SkyWcs> SkyWcs::readString(std::string& str) {
     std::istringstream is(str);
     return SkyWcs::readStream(is);
 }
 
-void SkyWcs::writeStream(std::ostream& os) const { detail::writeStream<SkyWcs>(*this, os); }
+void SkyWcs::writeStream(std::ostream& os) const {
+    os << SERIALIZATION_VERSION << " " << SkyWcs::getShortClassName() << " " << hasFitsApproximation() << " ";
+    getFrameDict()->show(os, false);  // false = do not write comments
+}
 
 std::string SkyWcs::writeString() const {
     std::ostringstream os;
     writeStream(os);
     return os.str();
 }
 
-SkyWcs::SkyWcs(std::shared_ptr<ast::FrameSet>&& frameSet) : TransformPoint2ToIcrsCoord(std::move(frameSet)){};
+std::string SkyWcs::getPersistenceName() const { return getSkyWcsPersistenceName(); }
```

Comment:
Because `afw::table::io::Persistable` requires the member function and SkyWcs.cc needs the same information without constructing a `SkyWcs` in order to register SkyWcsFactory. It is a common pattern using table persistence, and ince the free function is local to the .cc file, I don't see it enough of a problem to try to work around it.

Reply:
Ah, I missed that `getPersistenceName` wasn't static. Got it.